asteroid_filterbanks/pcen.py

```diff
@@ -0,0 +1,270 @@
+
+    def forward(
+        self, mag_spec: torch.Tensor, initial_state: Optional[torch.Tensor]
+    ) -> Tuple[torch.tensor, torch.Tensor]:
+        if initial_state is None:
+            initial_state = mag_spec[:, :, :, 0].unsqueeze(-1)
+
+        weights = torch.clamp(self.weights, 0.0, 1.0)
+        accumulator = initial_state.transpose(1, -1)
+        out = None
+        for x in torch.split(mag_spec.transpose(1, -1), 1, dim=1):
+            accumulator = weights * x + (1.0 - weights) * accumulator
+            accumulator = accumulator
+            if out is None:
+                out = accumulator.transpose(1, -1)
+            else:
+                out = torch.cat((out, accumulator.transpose(1, -1)), dim=-1)
+        return out, accumulator.transpose(1, -1)
+
+
+class TrainableParameters(TypedDict):
+    alpha: bool
+    delta: bool
+    root: bool
+    smooth: bool
+
+
+class _PCEN(nn.Module):
+    def __init__(
+        self,
+        alpha: float = 0.96,
+        delta: float = 2.0,
+        root: float = 2.0,
+        floor: float = 1e-6,
+        smooth: float = 0.04,
+        n_channels: int = 1,
+        trainable: Union[bool, TrainableParameters] = False,
+        per_channel_smoothing: bool = False,
+    ):
+        super().__init__()
+
+        if trainable is True or trainable is False:
+            trainable = TrainableParameters(
+                alpha=trainable, delta=trainable, root=trainable, smooth=trainable
+            )
+
+        self.trainable = trainable
+        self.n_channels = n_channels
+
+        self.alpha = nn.Parameter(
+            torch.full((self.n_channels,), fill_value=alpha), requires_grad=self.trainable["alpha"]
+        )
+        self.delta = nn.Parameter(
+            torch.full((self.n_channels,), fill_value=delta), requires_grad=self.trainable["delta"]
+        )
+        self.root = nn.Parameter(
+            torch.full((self.n_channels,), fill_value=root), requires_grad=self.trainable["root"]
+        )
+
+        self.floor = floor
+        self.ema = ExponentialMovingAverage(
+            smooth=smooth,
+            per_channel=per_channel_smoothing,
+            n_channels=self.n_channels,
+            trainable=self.trainable["smooth"],
+        )
+
+    def forward(
+        self, mag_spec: torch.Tensor, initial_state: Optional[torch.Tensor] = None
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if len(mag_spec.shape) == 3:
+            # If n_channels is 1, add a single dimension to keep the shape consistent with multichannel shapes.
+            mag_spec = mag_spec.unsqueeze(1)
+
+        alpha = torch.min(self.alpha, torch.tensor(1.0))
+        root = torch.max(self.root, torch.tensor(1.0))
+        one_over_root = 1.0 / root
+        ema_smoother, hidden_state = self.ema(mag_spec, initial_state)
+        # Equation (1) in [1]
+        out = (
+            mag_spec.transpose(1, -1) / (self.floor + ema_smoother.transpose(1, -1)) ** alpha
+            + self.delta
```

Comment:
Join lines

Reply:
This formatting is automatically applied by `black`.